docs/src/tutorials/getting_started/design_patterns_for_larger_models.jl

```diff
@@ -0,0 +1,610 @@
+# different. We've wrapped it in a function, defined some data types, and
+# introduced configuration options to control the variables and constraints that
+# get added. There's a few other steps we can do to further improve things. They
+# are:
+#  * remove the dependence on `GLPK`
+#  * add checks that we found an optimal solution
+#  * add a helper function to avoid the need to explicitly construct the data.
+
+function solve_knapsack_6(
+    optimizer,
+    data::KnapsackData,
+    config::AbstractKnapsack,
+)
+    model = Model(optimizer)
+    add_knapsack_variables(model, data, config)
+    add_knapsack_constraints(model, data, config)
+    add_knapsack_objective(model, data, config)
+    optimize!(model)
+    if termination_status(model) != OPTIMAL
+        @warn("Model not solved to optimality")
+        return nothing
+    end
+    return value.(model[:x])
+end
+
+function solve_knapsack_6(optimizer, data::String, config::AbstractKnapsack)
+    return solve_knapsack_6(optimizer, read_data(data), config)
+end
+
+data_filename = joinpath(@__DIR__, "data", "knapsack.json")
+solution = solve_knapsack_6(GLPK.Optimizer, data_filename, BinaryKnapsack())
+
+# ## Create a module
+
+# Now we're ready to expose our model to the wider-world. That might be as part
+# of a larger Julia project that we're contributing to, or as a stand-alone
+# script that we can run on-demand. In either case, it's good practice to wrap
+# everything in a module. This further encapsulates our code into a single
+# namespace, and we can add documentation in the form of docstrings.
+
+# Some good rules to follow whe creating a module are:
+# * Use `import` in a module instead of `using` to make it clear which functions
+#   are from which packages
+# * Use `_` to start function and type names that are considered private
+# * Add docstrings to all public variables and functions.
+
+module KnapsackModel
+
+import JuMP
+import JSON
+
+struct _KnapsackObject
+    profit::Float64
+    weight::Float64
+    function _KnapsackObject(profit::Float64, weight::Float64)
+        if weight < 0
+            throw(DomainError("Weight of object cannot be negative"))
+        end
+        return new(profit, weight)
+    end
+end
+
+struct _KnapsackData
+    objects::Dict{String,_KnapsackObject}
+    capacity::Float64
+end
+
+function _read_data(filename)
+    d = JSON.parsefile(filename)
+    return _KnapsackData(
+        Dict(
+            k => _KnapsackObject(v["profit"], v["weight"]) for
+            (k, v) in d["objects"]
+        ),
+        d["capacity"],
+    )
+end
+
+abstract type _AbstractKnapsack end
+
+"""
+    BinaryKnapsack()
+
+Create a binary knapsack problem where each object can be taken 0 or 1 times.
+"""
+struct BinaryKnapsack <: _AbstractKnapsack end
+
+"""
+    IntegerKnapsack()
+
+Create an integer knapsack problem where each object can be taken any amount of
+times.
+"""
+struct IntegerKnapsack <: _AbstractKnapsack end
```

Comment:
That's dedication to the `_x` is private rule by using on your module supertype!